### PR TITLE
As an operator I want to be able to change the KVM hypervisor credentials

### PR DESCRIPTION
**Pull request by wilderrodrigues**
CLOUDSTACK-8589

API call (updateHostPassword) changes credentials on XenServer hypervisor but not on KVM. This should be extended also for KVM

* Decouple the use of updateHostPassword
   - The code was hard to maintain because updating a host or all the hosts in a cluster was handled in the same method
   - Created updateHost and updateCluster password in both ResourceManager and ManagementServer interfaces/classes
   - The check for which method to use is done in the API level
   - Started adding the support for KVM host password update
* Fix a bug in the doUpdateHostPassword() method
  - 2 years ago a change in that method caused an issue when trying to update the hosts password: they would be updated in the database but not in the host it self, causing the hosts to disconnect after a management server restart.
* Making possible to update KVM hosts password in the database via the API
   - Make sure the doUpdateHostPassword() method returns the result from the answer, not always true
   - Added the LibvirtUpdateHostPasswordCommandWrapper class, which will handle the changes in the KVM agent that will be further implemented in the future. The command will avoid the agent to complain about Unsupported command
   - Added a test to make sure the current implementation of the comment works. If changes in the future, it will also require changes. In that way, we make sure nobody will break it

I already did basic manual tests with Management Server running on CentOS 7.1; 2 XenServer hosts; and 1 KVM host. All deployed in 2 advances zones. I was able to successfully execute:

Update KVM host
- update hostpassword hostid=95aa28ae-8038-4457-8951-504e760b53a4  username=root password=admin

Update 1 XenServer host
- update hostpassword hostid=9b236e34-0cc7-4d36-8b01-925fb10b4c25 username=root password=password

Update a XenServer cluster
- update hostpassword hostid=9b236e34-0cc7-4d36-8b01-925fb10b4c25 clusterid=fb58b84d-f9bb-4603-8b85-b56a49e1676f  username=root password=password

I also created isolated networks and virtual machines. The machines deployed on the XenServer zone also used affinity groups. I deployed 1 virtual machine in the KVM zone.

 @DaanHoogland @remibergsma @bhaisaab 

Please, do not merge yet. I will also run the integration/smoke tests.

In addition, all the commits are atomic. I also took care of splitting the commits related to "formatting code" and fixes.

Cheers,
Wilder

### CODE REVIEW

**Comment by terbolous** [comment]
Would it be too much to ask for a corresponding Jira ticket? Definitely makes it easier to track changes when writing release notes :-)

**Comment by wilderrodrigues** [comment]
Indeed. :)

I will create the ticket on ACS Jira and add to the  PR description. 

Cheers,
Wilder

**Comment by terbolous** [comment]
Excellent, thanks :-)

**Comment by bhaisaab** [comment]
@wilderrodrigues also squash them :) will this change passwords on the hosts?

**Comment by wilderrodrigues** [comment]
Hi @bhaisaab 

No, it doesn't change pwd in the hosts. I will implement that in a separate story/issue.

Concerning the squash, I won't do it. Why? The commits are atomic, which makes easier to review (e.g. one doesn't need t review commits that say "formatting code..."). I split code formatting from the fixes. Another point, if there is a change in a given commit, I will change just that commit and push it, which reduces the work for the person behind the fixes as well. On a third point, if we squash atomic commits, the reviewer will have the whole huge thing to look at, as everything has just be done together.

In short, squashing those atomic commits will bring no benefit at all.

Squash would only help if someone messes up with the commits and have everything tangled, which is not the case with this PR.

If you look at the bullet point on the PR description, those are the most relevant commits to look at. I even took care of using the same commit title for the bullet points.

I will start testing this evening.

Cheers,
Wilder

**Comment by bhaisaab** [comment]
@wilderrodrigues it's difficult to track changes, against JIRA which is why I asked for them to be squashed and use the CLOUDSTACK-8589. I think a single commit to capture the changes would be cleaner, but that just my aesthetics preferences. Please merge when you're done testing, thanks.

**Comment by wilderrodrigues** [comment]
Hi @bhaisaab 

Unfortunately I forgot to create the issue before the commits, otherwise it wouldn't be a problem now.

Since I added the issue number in the last commit, I hope it won't cause pain during 4.6 release. In any case, I will double the attention before the PRs and will also look closer to the 4.6 release and help with any thing the pops up under my name.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
XenServer 6.2 test report:

Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... === TestName: test_08_migrate_vm | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok
Test router internal advanced zone ... === TestName: test_02_router_internal_adv | Status : SUCCESS ===
ok
Test restart network ... === TestName: test_03_restart_network_cleanup | Status : SUCCESS ===
ok
Test router basic setup ... === TestName: test_05_router_basic | Status : SUCCESS ===
ok
Test router advanced setup ... === TestName: test_06_router_advanced | Status : SUCCESS ===
ok
Test stop router ... === TestName: test_07_stop_router | Status : SUCCESS ===
ok
Test start router ... === TestName: test_08_start_router | Status : SUCCESS ===
ok
Test reboot router ... === TestName: test_09_reboot_router | Status : SUCCESS ===
ok
Test to create service offering ... === TestName: test_01_create_service_offering | Status : SUCCESS ===
ok
Test to update existing service offering ... === TestName: test_02_edit_service_offering | Status : SUCCESS ===
ok
Test to delete service offering ... === TestName: test_03_delete_service_offering | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_remote_access_vpn | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_site2site_vpn | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 22 tests in 2634.005s

OK
/tmp//MarvinLogs/test_vpc_vpn_YESELF/results.txt (END)

* Updating single XenServer host:

update hostpassword hostid=2ffa8219-08cb-4014-bcb6-49f58ca0c9b6 username=root password=admin

2015-06-26 05:12:36,470 DEBUG [c.c.s.ManagementServerImpl] (2135942710@qtp-1568127297-46:ctx-e783efbc ctx-0852e7de ctx-9829cd2e) Changing password for host name = xen1
2015-06-26 05:12:36,499 DEBUG [c.c.a.t.Request] (2135942710@qtp-1568127297-46:ctx-e783efbc ctx-0852e7de ctx-9829cd2e) Seq 2-8034984685182386389: Sending  { Cmd , MgmtId: 3232241213, via: 2(xen1), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.UpdateHostPasswordCommand":{"
wait":0}}] }

* Updating XenServer cluster:

update hostpassword username=root password=admin clusterid=63b8ff8c-4d0c-4911-aad4-6c30a3257a29

2015-06-26 05:14:47,596 DEBUG [c.c.s.ManagementServerImpl] (2135942710@qtp-1568127297-46:ctx-5144239f ctx-36c6878b ctx-d371bef2) Changing password for host name = xen1
2015-06-26 05:14:47,599 DEBUG [c.c.s.ManagementServerImpl] (2135942710@qtp-1568127297-46:ctx-5144239f ctx-36c6878b ctx-d371bef2) Changing password for host name = xen2
2015-06-26 05:14:47,614 DEBUG [c.c.a.t.Request] (2135942710@qtp-1568127297-46:ctx-5144239f ctx-36c6878b ctx-d371bef2) Seq 2-8034984685182386394: Sending  { Cmd , MgmtId: 3232241213, via: 2(xen1), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.UpdateHostPasswordCommand":{"wait":0}}] }
2015-06-26 05:14:47,614 DEBUG [c.c.a.t.Request] (2135942710@qtp-1568127297-46:ctx-5144239f ctx-36c6878b ctx-d371bef2) Seq 2-8034984685182386394: Executing:  { Cmd , MgmtId: 3232241213, via: 2(xen1), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.UpdateHostPasswordCommand":{"wait":0}}] }

I will now test the KVM host

**Comment by runseb** [comment]
I was with @bhaisaab on this, I would have like to see a JIRA issue. We did not call it so yet, but we should consider being in master freeze right now, and only things with proper JIRA ticket should be merged. I would also prefer squashed commits...

**Comment by wilderrodrigues** [comment]
I created the issue and added to the last commit. Did it last week:

https://issues.apache.org/jira/browse/CLOUDSTACK-8589

Perhaps you missed that.

Concerning squashed commits, ew should really have discussed about it when you were here. I'm 100% if I squash my atomic commits next time you guys will complain. It seems I haven't been clear enough about that yet.

@remibergsma @miguelaferreira Would you guys give your input on these "squash commits" matter? I'm a bit tired of this discussion. For me, squashed commits won't make a review easier, unless there is no review at all or if the committer has changes a couple of lines in a bash script.

What I would like to see is people testing changes the way I do.

Cheers,
Wilder

**Comment by miguelaferreira** [comment]
As long as people make good commits, that are atomic, capture relevant changes together, and have descriptive messages, IMHO squashing only make things worse.

If you squash commits, all you get is less commits. Some people might say a cleaner master branch.
I won't discuss aesthetics here. I will try to be objective.

With less commits you also get larger commits. That means that in a change set (e.g. a PR) where a large number of changes are made, these changes are all bundled together. That forces the reviewer to take all the context at once. 
With more commits you get smaller commits. That allows the reviewer to take chunks of context at a time. And, in addition, if the reviewer wants/needs to get all the context at once, it is easy to make a diff between the first and the last commit in the change set (Github offers both per commit view and total diff view, just see the tabs on top of this PR).

If you will, this is comparable to reviewing a class will several small methods or reviewing a class with one big method.

Another interesting aspect that should be considered is reverting commits. As you may all know this can happen quite frequently. With smaller, atomic commits, it is easer to revert a change (e.g a bug introducing change) in a particular part of the code (while leaving the rest of the code untouched). It is easier because (1) the small commit isolates the change; (2) in the presence of conflicts these can be fixed iteratively, in small steps.

Finally, I would like to point out that smaller commits allow the developer to convey more information about each change. If you squash a set of commits you loose the description of all the small changes.
Do we really want to go towards less information about code changes? Or do we want to have the developers write a bullet list of the changes they made and assign them to parts of the code? There simply isn't any other way to preserve this information.

Of course that, as I already said, all this only makes sense if:
* commits are small;
* atomic;
* have good descriptions.

And that is, simply put, entry-level development discipline.
